### PR TITLE
[BEAM-2762] Python code coverage report in Postcommit

### PR DESCRIPTION
**Pull request by markflyhigh**
Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] Make sure there is a [JIRA issue](https://issues.apache.org/jira/projects/BEAM/issues/) filed for the change (usually before you start working on it).  Trivial changes like typos do not require a JIRA issue.  Your pull request should address just this issue, without pulling in other changes.
 - [ ] Each commit in the pull request should have a meaningful subject line and body.
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue.
 - [ ] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
 - [ ] Run `mvn clean verify` to make sure basic checks pass. A more thorough check will be performed on your pull request automatically.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

---

Use coverage.py to run py27gcp test suite in order to generate coverage report in Python postcommit build.
 - Add `cover` env in `tox.ini` which run `py27gcp` and generate report.
 - Report will be printed to test log and also in a xml file.
 - Only enable the coverage report in python postcommit since precommit build already takes ~30mins with existing test suites. Run `py27gcp` in code coverage will add ~7mins to the build. (thoughts?)

Example report: [https://builds.apache.org/job/beam_PostCommit_Python_Verify/3183/consoleFull](https://builds.apache.org/job/beam_PostCommit_Python_Verify/3183/consoleFull)
```
Name                                                                              Stmts   Miss Branch BrPart  Cover
-------------------------------------------------------------------------------------------------------------------
apache_beam/__init__.py                                                              12      1      2      1    86%
apache_beam/coders/coder_impl.py                                                    411     17    100      9    95%
...
apache_beam/utils/timestamp.py                                                       82      3     14      0    97%
apache_beam/utils/windowed_value.py                                                  38      2      6      0    95%
-------------------------------------------------------------------------------------------------------------------
TOTAL                                                                             34291   3655   6471    689    88%
```

Future:
Send coverage report to coverage service (like `coveralls.io`). The known issue for `coveralls.io` is that reports from different languages will mix together in `coveralls.io` and record as history data.

+R: @robertwb

### CODE REVIEW

**Comment by markflyhigh** [comment]
Run Python PostCommit

**Comment by markflyhigh** [comment]
friendly ping @robertwb

File: sdks/python/pom.xml [hunk 1]
```diff
@@ -200,7 +200,7 @@
                   <executable>${python.user.base}/bin/tox</executable>
                   <arguments>
                     <argument>-e</argument>
-                    <argument>ALL</argument>
+                    <argument>py27,py27gcp,py27cython,lint,docs</argument>
```
Comment (robertwb) [inline sdks/python/pom.xml]: Why this change?

File: sdks/python/setup.cfg [hunk 2]
```diff
@@ -25,3 +25,30 @@ verbosity=2
...
+
+# Configuration for code coverage.
+# .coveragerc to control coverage.py
+[coverage:run]
+branch = True
+source = apache_beam
+
+[coverage:report]
+# Regexes for lines to exclude from consideration
+exclude_lines =
```
Comment (robertwb) [inline sdks/python/setup.cfg]: Maybe exclude the auto-generated files in portability/api?

File: sdks/python/pom.xml (hunk 1 above)
Comment (robertwb) [reply sdks/python/pom.xml]: Is this so precommit excludes coverage?

Comment (markflyhigh) [reply sdks/python/pom.xml]: yes, coverage tool will run full set of `py27gcp` tests again which I think is unnecessary to include in Precommit considering current build time (1+hour).

File: sdks/python/setup.cfg (hunk 2 above)
Comment (markflyhigh) [reply sdks/python/setup.cfg]: According to the comment in files, I guess we can exclude all `*_pb2.py` and `pb2_grpc.py`?

**Comment by markflyhigh** [comment]
PTAL @robertwb

File: sdks/python/pom.xml (hunk 1 above)
Comment (robertwb) [reply sdks/python/pom.xml]: Is there any way to phrase this as a negative, so we get any possible new configurations by default? (If not, drop a comment that this is everything but coverage...)

File: sdks/python/setup.cfg [hunk 3]
```diff
@@ -31,24 +31,28 @@ exclude=fast_coders_test|typecoders_test
...
-    if self\.debug
+  # Don't complain about missing debug-only code:
+  def __repr__
+  if self\.debug
 
-    # Don't complain if tests don't hit defensive assertion code:
-    raise AssertionError
-    raise NotImplementedError
+  # Don't complain if tests don't hit defensive assertion code:
+  raise AssertionError
```
Comment (robertwb) [inline sdks/python/setup.cfg]: Do we every manually raise this?

File: sdks/python/pom.xml (hunk 1 above)
Comment (markflyhigh) [reply sdks/python/pom.xml]: Found a work around way: `tox envlist` will run everything defined in [tox:envlist] so we can exclude `cover` from it. As long as new defined configuration is added to [tox:envlist], maven build will include it.

File: sdks/python/setup.cfg (hunk 3 above)
Comment (markflyhigh) [reply sdks/python/setup.cfg]: Thanks for pointing it out. we don't raise it.

**Comment by markflyhigh** [comment]
PTAL @robertwb

**Comment by robertwb** [comment]
LGTM, thanks.

**Comment by markflyhigh** [comment]
retest this please

**Comment by markflyhigh** [comment]
retest this please

**Comment by markflyhigh** [comment]
Rebase this branch with master and retest.

**Comment by aaltay** [comment]
@markflyhigh is this ready to be merged?

**Comment by markflyhigh** [comment]
@aaltay Code is ready to merge but I think the submission is blocking on test with unrelated failure. I didn't got all pass after several times try. 

Let me rebase since it's been 10+ days old.

**Comment by aaltay** [comment]
@markflyhigh is this ready to be merged now?

**Comment by aaltay** [comment]
any updates?

**Comment by markflyhigh** [comment]
This PR is rebased against origin/master and ready to review now.

File: sdks/python/pom.xml [hunk 4]
```diff
@@ -226,8 +226,8 @@
                 <configuration>
                   <executable>${python.user.base}/bin/tox</executable>
                   <arguments>
-                    <argument>-e</argument>
-                    <argument>ALL</argument>
+                    <!--Testenv that are only defined in envlist will be triggered.-->
+                    <argument>envlist</argument>
```
Comment (aaltay) [inline sdks/python/pom.xml]: Where do we set the value of `envlist`?

File: sdks/python/setup.cfg [hunk 5]
```diff
@@ -25,3 +25,33 @@ verbosity=2
 # fast_coders module which is not available when running unit tests:
 # fast_coders_test and typecoders_test.
 exclude=fast_coders_test|typecoders_test
+
+# Configuration for code coverage.
+# .coveragerc to control coverage.py
```
Comment (aaltay) [inline sdks/python/setup.cfg]: Is `.coveragerc` a separate file? Or is this section the only configuration?

File: sdks/python/pom.xml (hunk 4 above)
Comment (markflyhigh) [reply sdks/python/pom.xml]: cover will be triggered if running `tox -e ALL`. Since I didn't add cover to envlist in tox.ini, Precommit will not run cover in this case. 

This approach has two cons:
1. cover is not belong to default tox testenv.
2. testenv changes will not reflect in Precommit if they are not defined in envlist.

I tried to find a tox command without envlist to exclude cover in execution but failed.

File: sdks/python/setup.cfg (hunk 5 above)
Comment (markflyhigh) [reply sdks/python/setup.cfg]: `.coveragerc` is the following section to config coverage.py. Looks like we should remove it to avoid confusion.

**Comment by markflyhigh** [comment]
Thank you @aaltay. I tested it locally after latest rebase. I can also trigger postcommit here to verify.

**Comment by markflyhigh** [comment]
Run Python PostCommit

File: sdks/python/pom.xml (hunk 4 above)
Comment (aaltay) [reply sdks/python/pom.xml]: How long it takes to run coverage jobs?

Could you add a comment here and `tox.ini` that new environments will be excluded by default unless explicitly added to `envlist`.

File: sdks/python/setup.cfg (hunk 5 above)
Comment (aaltay) [reply sdks/python/setup.cfg]: Do you mean removing this line of comment?

Comment (markflyhigh) [reply sdks/python/setup.cfg]: yes, ling 30, if it makes confusion. what do you think?

File: sdks/python/pom.xml (hunk 4 above)
Comment (markflyhigh) [reply sdks/python/pom.xml]: will do.

according to [this Postcommit run](https://builds.apache.org/job/beam_PostCommit_Python_Verify/4008/), `cover` ran 1488 tests in 245.447s and `py27gcp` ran same test set in 158.642s.

Comment (aaltay) [reply sdks/python/pom.xml]: OK. Let's keep it separate.

File: sdks/python/setup.cfg (hunk 5 above)
Comment (aaltay) [reply sdks/python/setup.cfg]: OK.

**Comment by aaltay** [comment]
LGTM, ping me when you make the changes.

**Comment by markflyhigh** [comment]
@aaltay done. PTAL

**Comment by aaltay** [comment]
LGTM. I will merge after tests pass.

**Comment by robertwb** [comment]
For future reference, we should probably squash commits starting with fixup! or similar before merging.

**Comment by aaltay** [comment]
Thank you Robert. Will squash or ask the author to re-organize commits next time.
